packages/charts/react-charting/src/components/ResponsiveContainer/ResponsiveContainer.tsx

```diff
@@ -0,0 +1,73 @@
+import * as React from 'react';
+import { classNamesFunction, getWindow } from '@fluentui/react';
+import {
+  IResponsiveChildProps,
+  IResponsiveContainerProps,
+  IResponsiveContainerStyles,
+} from './ResponsiveContainer.types';
+import { getStyles } from './ResponsiveContainer.styles';
+
+const getClassNames = classNamesFunction<{}, IResponsiveContainerStyles>();
+
+/**
+ * Responsive Container component
+ * {@docCategory ResponsiveContainer}
+ */
+export const ResponsiveContainer: React.FC<IResponsiveContainerProps> = props => {
+  const containerRef = React.useRef<HTMLDivElement>(null);
+  const onResizeRef = React.useRef<IResponsiveContainerProps['onResize']>();
+  const classNames = React.useMemo(() => getClassNames(getStyles()), []);
+
+  const [size, setSize] = React.useState<{ containerWidth?: number; containerHeight?: number }>({});
+
+  onResizeRef.current = props.onResize;
+
+  React.useEffect(() => {
+    const _window = getWindow(containerRef.current) as (Window & typeof globalThis) | undefined;
+    let animationFrameId: number | undefined;
+    let resizeObserver: ResizeObserver | undefined;
+
+    const resizeCallback = (entries: ResizeObserverEntry[]) => {
+      const { width: containerWidth, height: containerHeight } = entries[0].contentRect;
+      // rAF is an alternative to the throttle function. For more info, see:
+      // https://css-tricks.com/debouncing-throttling-explained-examples/#aa-requestanimationframe-raf
+      animationFrameId = _window?.requestAnimationFrame(() => {
+        setSize(prevSize => {
+          const roundedWidth = Math.floor(containerWidth);
+          const roundedHeight = Math.floor(containerHeight);
+          if (prevSize.containerWidth === roundedWidth && prevSize.containerHeight === roundedHeight) {
+            return prevSize;
+          }
+
+          return { containerWidth: roundedWidth, containerHeight: roundedHeight };
+        });
+      });
+      onResizeRef.current?.(containerWidth, containerHeight);
+    };
+
+    if (_window && _window.ResizeObserver) {
+      resizeObserver = new _window.ResizeObserver(resizeCallback);
+      if (containerRef.current) {
+        resizeObserver.observe(containerRef.current);
+      }
+    }
+
+    return () => {
+      if (animationFrameId) {
+        _window?.cancelAnimationFrame(animationFrameId);
+      }
+
+      resizeObserver?.disconnect();
+    };
+  }, []);
+
+  return (
+    <div ref={containerRef} className={classNames.root} style={{ width: props.width, height: props.height }}>
+      {React.cloneElement<IResponsiveChildProps>(props.children, {
```

Comment:
>cloneElement [](http://example.com/codeflow?start=13&length=12)

how expensive is cloneElement.

Reply:
cloneElement is basically React.createElement. It returns a new element when props passed to it changes. It can have performance implications when used excessively or incorrectly, but it seems appropriate in this case.

Alternatives like using a render prop or context for data passing are either less user-friendly or would require more changes in the chart implementation.

References:
[javascript - React.cloneElement in List performance - Stack Overflow](https://stackoverflow.com/questions/54922160/react-cloneelement-in-list-performance)
[`React.cloneElement` vs. `Render Prop` pattern benchmarking](https://gist.github.com/nemoDreamer/21412b28dc65d51e2c5c8561a8f82ce1)